### PR TITLE
Added feature for deleting a user

### PR DESCRIPTION
**Pull request by ibarrae**

### CODE REVIEW

File: src/User/Persistence.hs [hunk 1]
```diff
@@ -9,4 +10,7 @@ selectAllUsers :: (HasPostgres m) => m [User]
 selectAllUsers = query_ allUsersQuery
 
 insertUser :: (HasPostgres m) => UserForm -> m Int64
-insertUser = execute insertUserQuery
\ No newline at end of file
+insertUser = execute insertUserQuery
+
+deleteUser :: (HasPostgres m) => Int -> m Int64
+deleteUser key = execute deleteUserQuery $ toRow [key]
```
Comment (CristhianMotoche) [inline src/User/Persistence.hs]: @kinorius123 Is the `toRow` function necessary?

Comment (ibarrae) [reply src/User/Persistence.hs]: @CristhianMotoche You are right, removing it right now.

File: test/SiteSpec.hs [hunk 2]
```diff
@@ -22,6 +22,14 @@ spec =
...
         ci <- loadConnInfoFromEnv
-        er <- T.runHandler Nothing (get "/users/add" M.empty) handleUsers $ appInit ci
+        er <- T.runHandler Nothing (get "/users/add" M.empty) handleUserAdd $ appInit ci
+        let r = U.fromRight er
+        rspStatus r `shouldBe` 200
+    context "when requesting /users/delete/:id" $
+      it "should respond with status 200" $ do
+        ci <- loadConnInfoFromEnv
+        er <- T.runHandler Nothing 
+          (delete "/users/delete/id" $ M.fromList [("id", ["500"])])
```
Comment (CristhianMotoche) [inline test/SiteSpec.hs:64]: Why `500`? @kinorius123 
I think you need to create the user before delete it.

File: src/Routes.hs [hunk 3]
```diff
@@ -1,14 +1,17 @@
...
 import Application
 import User.Handlers
 
 routes :: [(ByteString,AppHandler)]
-routes = [ ("/", welcomePage)
-         , ("/users", handleUsers)
-         , ("/users/add", handleUserAdd)]
+routes = [ ("/users", handleUsers)
+         , ("/users/add", handleUserAdd)
+         , ("/users/delete/:id", method DELETE handleUserDelete)
```
Comment (CristhianMotoche) [inline src/Routes.hs]: @kinorius123 Can we also specify the `method`s for the other handlers?

File: package.yaml [hunk 4]
```diff
@@ -74,3 +74,6 @@ tests:
     - hspec
     - snap
     - containers
+    - hspec-pg-transact
+    - pg-transact
```
Comment (CristhianMotoche) [inline package.yaml]: @kinorius123 Where are you using this libraries?

File: .gitignore [hunk 5]
```diff
@@ -4,4 +4,5 @@ log/
 dist/
 .ghci
 .env
+UserQueriesSpec.hs.old
```
Comment (CristhianMotoche) [inline .gitignore:7]: @kinorius123 Is necessary to add this file to the `.gitignore`? If this file is only generated in your computer, could you please only delete it?

File: ChangeLog.md [hunk 6]
```diff
@@ -1,5 +1,8 @@
 # Changelog for snap
-
+## 1.0.6
+    - Added the feature of deleting a user with a DELETE request.
+    - Added tests for user queries.
```
Comment (CristhianMotoche) [inline ChangeLog.md:4]: @kinorius123 I didn't find any test specifically for the queries.

File: snap-example.cabal [hunk 7]
```diff
@@ -110,6 +109,7 @@ test-suite snap-example-test
   other-modules:
       SiteSpec
       UserPresenterSpec
+      UserQueriesSpec
```
Comment (CristhianMotoche) [inline snap-example.cabal]: @kinorius123 Where is this module?

Comment (CristhianMotoche) [reply snap-example.cabal]: @kinorius123 Did you push your latest changes?

File: src/User/Handlers.hs [hunk 8]
```diff
@@ -22,5 +24,18 @@ handleUserAdd = do
   sd <- liftIO getSystemDay
   (view, muf) <- runForm "user_form" $ userForm sd
   case muf of
-    Just uf -> insertUser uf >> redirect "/users"
-    Nothing -> heistLocal (bindDigestiveSplices view) $ render "users_add" 
\ No newline at end of file
+    Just uf -> do
+      r <- insertUser uf
```
Comment (CristhianMotoche) [inline src/User/Handlers.hs]: @kinorius123 Could you please rename this to something like `numCreatedUsers`? We do prefer explicit names.

File: src/User/Handlers.hs [hunk 9]
```diff
@@ -22,5 +24,18 @@ handleUserAdd = do
...
+        (redirect "/users") (r>0)
+    Nothing -> heistLocal (bindDigestiveSplices view) $ render "users_add" 
+
+handleUserDelete :: AppHandler
+handleUserDelete = do
+  key <- getParam "id"
+  r <- deleteUser 
+        (maybe 
+          (error "Could not parse id parameter") 
+          (\bs -> read $ BS.unpack bs :: Int) key)
```
Comment (CristhianMotoche) [inline src/User/Handlers.hs]: @kinorius123 Can you replace `\bs -> read $ BS.unpack bs` with `read . BS.unpack`?
@kinorius123 Also, could you please use a more explicit name than `r`?

File: src/User/Handlers.hs [hunk 10]
```diff
@@ -22,5 +24,18 @@ handleUserAdd = do
...
+      bool (writeBS "Something went wrong while adding the user.") 
+        (redirect "/users") (r>0)
+    Nothing -> heistLocal (bindDigestiveSplices view) $ render "users_add" 
+
+handleUserDelete :: AppHandler
+handleUserDelete = do
+  key <- getParam "id"
+  r <- deleteUser 
+        (maybe 
+          (error "Could not parse id parameter")
```
Comment (CristhianMotoche) [inline src/User/Handlers.hs]: @kinorius123 What about using [`serverError`](http://hackage.haskell.org/package/snap-extras-0.12.0.0/docs/Snap-Extras-CoreUtils.html) instead on `error`?

File: src/User/Handlers.hs [hunk 11]
```diff
@@ -22,5 +24,18 @@ handleUserAdd = do
...
+    Nothing -> heistLocal (bindDigestiveSplices view) $ render "users_add" 
+
+handleUserDelete :: AppHandler
+handleUserDelete = do
+  key <- getParam "id"
+  r <- deleteUser 
+        (maybe 
+          (error "Could not parse id parameter") 
+          (\bs -> read $ BS.unpack bs :: Int) key)
+  bool (writeBS "Something went wrong while deleting the user.")
```
Comment (CristhianMotoche) [inline src/User/Handlers.hs]: I think a `serverError` will work well here, too. WDYT? @kinorius123

File: snap-example.cabal [hunk 12]
```diff
@@ -2,10 +2,10 @@
 --
 -- see: https://github.com/sol/hpack
 --
--- hash: c7f0e45cb6b400590c94547ad8f80bd079187895126043ce8f84f9a6b2af8a6a
+-- hash: 908cb6b71c9871d5ef7a55febb823b0c357cfdcb09d0178810b156e3b01596c9
 
 name:           snap-example
-version:        1.0.5
+version:        1.0.6
```
Comment (dbalseiro) [inline snap-example.cabal]: About versioning, check these two articles:
https://semver.org/
https://blog.codinghorror.com/whats-in-a-version-number-anyway/

**Review by dbalseiro — COMMENTED** [review]
Good job @kinorius123 
Excellent review @CristhianMotoche thanks!

File: snap-example.cabal (hunk 12 above)
Comment (ibarrae) [reply snap-example.cabal]: Thanks @dbalseiro, gonna read those links right away.

File: test/UserQueriesSpec.hs [hunk 13]
```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE ScopedTypeVariables #-}
```
Comment (CristhianMotoche) [inline test/UserQueriesSpec.hs]: Good job @kinorius123 :+1:

**Pull request merged**